### PR TITLE
Throttle based on lag of all slaves in replication chain

### PR DESCRIPTION
**Pull request by jordanwheeler**
This is my slave lag throttler code as compared to soundcloud/lhm master (the `lhm-upstream` branch in this repo).  I've tested it quite a bit with Shopify.  There's some small changes we need to make there to use this (https://github.com/Shopify/shopify/pull/49019), but other than the `logger_params` change I needed to make in that PR, upstream lhm works just fine for us.

As I've mentioned, I'm not PRing this upstream because it explicitly uses the `Mysql2::Client`, whereas upstream "is compatible and continuously tested with MRI 2.0.x, 2.1.x, ActiveRecord 3.2.x and 4.x (mysql and mysql2 adapters)."  But we're more concerned with throttling slave lag than we are with getting this code upstream right now.

So I think the plan right now (pending review) is to merge the `lhm-upstream` branch, merge this, merge my Shopify PR (along with an updated gem ref) then start throttling some slave lag.

@camilo @arthurnn @jasonhl @shuhaowu

### CODE REVIEW

File: lib/lhm/throttler/slave_lag.rb [hunk 1]
```diff
@@ -43,55 +45,83 @@ def throttle_seconds
...
-        ActiveRecord::Base.send(adapter_method, config)
+      def lag
+        query_connection(SQL_SELECT_MAX_SLAVE_LAG, 'Seconds_Behind_Master')
       end
 
-      # This method fetch the Seconds_Behind_Master, when exec_query is no available, on AR 2.3.
-      def fetch_slave_seconds(result)
-        unless result.is_a? Mysql::Result
-          Lhm.logger.info "Not a Mysql::Result from the slave assuming 0 lag"
-          return 0
```
Comment (jordanwheeler) [inline lib/lhm/throttler/slave_lag.rb]: This is why I though soundcloud/lhm still supported ActiveRecord 2.x FYI.

File: lib/lhm/throttler/slave_lag.rb [hunk 2]
```diff
@@ -43,55 +45,83 @@ def throttle_seconds
...
+        rescue Mysql2::Error => e
+          Lhm.logger.info "Error conecting to #{@host}: #{e}"
+          nil
         end
+      end
 
-        keys = []
-        result.each_hash { |h| keys << h['Seconds_Behind_Master'].to_i }
-        keys
+      def config(get_config)
```
Comment (shuhaowu) [inline lib/lhm/throttler/slave_lag.rb]: Does passing in a config hash to LHM makes more sense than passing in a connection? It seems like that the consumer of this API should already have a connection object handy and that might be more appropriate to pass in rather than have this class create a new connection object?

Comment (jordanwheeler) [inline lib/lhm/throttler/slave_lag.rb]: It has a connection to the master, but does not have a connection to the slaves in the replication chain.  A connection to the slaves is what's being created here.  

The `username`, `password`, and `database` are the same across the master/all slaves in a shard, so the configuration is passed in and a new connection is created based on that.

Comment (shuhaowu) [inline lib/lhm/throttler/slave_lag.rb]: Another question is why are we filtering out only the host/username/password/db for the new config? Shouldn't the entire thing be used instead with the host changed?

Comment (jordanwheeler) [inline lib/lhm/throttler/slave_lag.rb]: Maybe, yeah.  Most of the stuff in our config will simply be ignored by `Mysql2::Client`, but there's probably no harm in just passing the whole thing along.

File: lib/lhm/throttler/slave_lag.rb [hunk 3]
```diff
@@ -43,55 +45,83 @@ def throttle_seconds
...
-        unless result.is_a? Mysql::Result
-          Lhm.logger.info "Not a Mysql::Result from the slave assuming 0 lag"
-          return 0
+      private
+
+      def client(config)
+        begin
+          Mysql2::Client.new(config)
+        rescue Mysql2::Error => e
+          Lhm.logger.info "Error conecting to #{@host}: #{e}"
```
Comment (shuhaowu) [inline lib/lhm/throttler/slave_lag.rb]: `s/conecting/connecting`

File: lib/lhm/throttler/slave_lag.rb [hunk 4]
```diff
@@ -43,55 +45,83 @@ def throttle_seconds
...
-        @connection.select_values(SQL_SELECT_SLAVE_HOSTS)
+        slaves = []
+        slave_hosts = master_slave_hosts
+        while slave_hosts.any? do
+          host = slave_hosts.pop
+          slave = Slave.new(host, @get_config)
+          if slaves.map(&:host).exclude?(host) && slave.connection
+            slaves << slave
+            slave_hosts << slave.slave_hosts
+            slave_hosts.flatten!
```
Comment (shuhaowu) [inline lib/lhm/throttler/slave_lag.rb]: `slave_hosts.concat(slave.slave_hosts)` is clearer?

**Comment by camilo** [comment]
paging @arthurnn

**Comment by jordanwheeler** [comment]
I'm going to :ship: this [Shopify PR](https://github.com/Shopify/shopify/pull/49019) which uses [this LHM ref](https://github.com/camilo/lhm/compare/master...Shopify:slave-lag-rebased) which consists of this PR rebased onto soundcloud/lhm master.

File: lib/lhm/throttler/slave_lag.rb [hunk 5]
```diff
@@ -43,55 +45,78 @@ def throttle_seconds
...
+          slave = Slave.new(host, @get_config)
+          if slaves.map(&:host).exclude?(host) && slave.connection
+            slaves << slave
+            slave_hosts.concat(slave.slave_hosts)
+          end
+        end
+        slaves
+      end
+
+      def master_slave_hosts
```
Comment (camilo) [inline lib/lhm/throttler/slave_lag.rb]: this is an interesting name for this function.

Comment (jordanwheeler) [inline lib/lhm/throttler/slave_lag.rb]: yeah, i honestly struggled with the naming of a few of these...

it's selecting slave hosts from the master :confused:

File: lib/lhm/throttler/slave_lag.rb [hunk 6]
```diff
@@ -43,55 +45,78 @@ def throttle_seconds
...
       def get_slaves
-        @connection.select_values(SQL_SELECT_SLAVE_HOSTS)
+        slaves = []
+        slave_hosts = master_slave_hosts
+        while slave_hosts.any? do
+          host = slave_hosts.pop
+          slave = Slave.new(host, @get_config)
+          if slaves.map(&:host).exclude?(host) && slave.connection
+            slaves << slave
+            slave_hosts.concat(slave.slave_hosts)
```
Comment (camilo) [inline lib/lhm/throttler/slave_lag.rb]: why is this needed ?

File: lib/lhm/throttler/slave_lag.rb [hunk 7]
```diff
@@ -1,5 +1,11 @@
 module Lhm
   module Throttler
+
+    def self.format_hosts(hosts)
+      hosts.map { |host| host.partition(':')[0] }
```
Comment (camilo) [inline lib/lhm/throttler/slave_lag.rb]: interesting so we are assuming always standard port?

File: lib/lhm/invoker.rb [hunk 8]
```diff
@@ -68,7 +68,8 @@ def normalize_options(options)
       end
 
       if options[:throttler]
-        options[:throttler] = Throttler::Factory.create_throttler(options[:throttler])
+        throttler_options = options[:throttler_options] || {}
+        options[:throttler] = Throttler::Factory.create_throttler(options[:throttler], throttler_options)
```
Comment (camilo) [inline lib/lhm/invoker.rb]: I like this

File: lib/lhm/throttler/slave_lag.rb (hunk 6 above)
Comment (jordanwheeler) [inline lib/lhm/throttler/slave_lag.rb]: we have a connection to the master, and we need to get the slaves for that master, then connect to each of those slaves, then get those slaves' slaves, etc. on down the chain.

`slaves` is an array of the slave objects that have their replication lag checked, which is what this method actually returns.  `slave_hosts` is an array consisting of all IP addresses slaving off of anything in the chain.  this `while` loop pops off each IP address in the `slave_hosts` array and creates a `Slave` object from that IP address.

so for example, when we get far enough down the chain this will try to connect to a hadoop instance, since that instance is slaving off the reader in a shard.  calling `slave.slave_hosts` on the reader will return the hadoop node's IP, that gets added to the `slave_hosts` array, the `while slave_hosts.any?` loop will eventually try to connect to that instance and fail, so `slave.connection` will be `nil`, this `if` condition will not be met and that slave will not be added to our final array of slaves that we want to check for slave lag.

so TL;DR the `slave_hosts` array exists to keep track of all slaves in the chain, and `slaves` is a separate array of slaves that we actually want to check.

File: lib/lhm/throttler/slave_lag.rb (hunk 7 above)
Comment (jordanwheeler) [inline lib/lhm/throttler/slave_lag.rb]: yeah.  i actually just moved this code around so it could be shared, it was already part of `SlaveLag`.

**Comment by arthurnn** [comment]
Before I dig in, what was the main reason the first code was not working? what was the issues with it? and why we had to change it so much?

**Comment by jordanwheeler** [comment]
There were three main problems we were having with the original code:

1) The original code wouldn't connect to slaves because the [config returned](https://github.com/soundcloud/lhm/blob/master/lib/lhm/throttler/slave_lag.rb#L78) would try and connect to the `shopify` database, and that doesn't work on other shards where the database name is `shopify_shard_#{shard_id}`.  For each shard `x` that an LHM runs on, we need to get the config with the database name `shopify_shard_x`

2) The original code selects the slaves from the master DB, but we've currently got 6 DBs in each shard's replication chain.  The original code would've checked the lag on the two slaves directly slaving off the master, but this will now walk the chain and check the lag on all of them.

3) Using `mysql_connection` threw an exception in Shopify core [here](https://github.com/Shopify/shopify/blob/master/lib/sharding/core_ext/active_record.rb#L45) because `pool` is nil.  I used `Mysql2::Client` instead to create the new connections to the slaves so we could avoid that; the slave connections don't care about ActiveRecord.

This goes along with [this Shopify PR](https://github.com/Shopify/shopify/pull/49019/files) which lets us get the new config on each shard.

**Comment by camilo** [comment]
hmm we have not pushed this upstream yet ? @jordanwheeler do you remember?
